usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -1372,6 +1435,34 @@ zil_lwb_write_issue(zilog_t *zilog, lwb_t *lwb)
 	lwb->lwb_issued_timestamp = gethrtime();
 	lwb->lwb_state = LWB_STATE_ISSUED;
 
+	/*
+	 * If this lwb does not have any threads waiting for it to
+	 * complete, we want to defer issuing the DKIOCFLUSHWRITECACHE
+	 * command to the vdevs written to by "this" lwb, and instead
+	 * rely on the "next" lwb to handle the DKIOCFLUSHWRITECACHE
+	 * command for those vdevs. Thus, we merge the vdev tree of
+	 * "this" lwb with the vdev tree of the "next" lwb in the list,
+	 * and assume the "next" lwb will handle flushing the vdevs (or
+	 * deferring the flush(s) again).
+	 *
+	 * This is a necessary performance optimization, especially for
+	 * workloads with lots of async write activity and few sync
+	 * write and/or fsync activity, as it has the potential to
+	 * coalesce multiple flush commands to a vdev into one.
+	 */
+	boolean_t defer_flush =
+	    (list_head(&lwb->lwb_waiters) == NULL && nlwb != NULL);
+
+	if (defer_flush) {
+		ASSERT3S(nlwb->lwb_state, ==, LWB_STATE_CLOSED);
+		while ((zv = avl_destroy_nodes(t, &cookie)) != NULL) {
+			vdev_t *vd = vdev_lookup_top(spa, zv->zv_vdev);
+			if (vd != NULL)
+				zil_lwb_add_vdev(nlwb, vd);
```

Comment:
I'm not sure I completely follow this comment..

Are you saying that, instead of freeing the node on the "current" lwb's tree, and potentially allocating a new node to insert into the "next" lwb's tree, I should just re-use the "current" lwb's node and avoid the extra alloc/free calls?

If I'm understanding you, that seems reasonable.. I'll look into doing it that way.

Reply:
Yeah, exactly.  Since one of the conditions of flush deferment is that `nlwb` is in `STATE_CLOSED`, we know that the entries in `lwb->lwb_vdev_tree` could simply be moved into `nlwb->lwb_vdev_tree`.  There should be no conflicts and no need to free and then reallocate all those nodes.